### PR TITLE
Aaron/collisionsensorfix

### LINKED ISSUE #7025: Possible Bug--Collision Sensor failed to report collision event in several scenarios
CARLA version: 0.9.13

Platform/OS: ubuntu 18.04 

Problem you have experienced: 
Possible Bug: The collision sensor doesn't work in some conditions.
What you expected to happen: The collision sensor should report the collision event whenever the collision happens.

Steps to reproduce: 
We used Scenario Runner and tested different scenarios like FollowLeadingVehicle, IntersectionCollisionAvoidance, LaneChangeSimple, etc. We discovered that sometimes collisions were visible to the naked eye, but the collision detector did not report the event. Here are some videos and the corresponding parameters:

https://github.com/carla-simulator/carla/assets/38175747/21aa04b3-c1a8-49a5-b3d6-0ccaa3cb26b1

Above is the FollowLeadingVehicle scenario, we set the initial driving speed of the vehicle is 20, which does not affect the final result because it seems that the vehicle doesn't need the acceleration process when changing speed in Carla. When the ego vehicle is 4m away from the NPC actor,  we change the driving speed to 6, and change the driving angle to (x = 1 and y = -0.31).

initial speed: 20
distance to change driving behavior: 4m 
speed after change driving behavior: 6
driving angle after change driving behavior: (x = 1, y = -0.31)

The collision sensor attached on ego vehicle doesn't report the collision event. For the sake of caution and fairness, we also attached a collision sensor on the NPC actor, it also failed to report the incident. We have run multiple rounds of tests in different scenarios and this problem appeared under different parameters. We are working on finding some patterns in the data results. Now at first glance, it looks like low speed, pedestrians, and bicycles are a few of the elements more prone to this problem, however, there are also cases of non-report when NPC is a vehicle. We will report the conclusion later.

### PR DESCRIPTION
**Pull request by Blyron**
#### Description

Exposing collision sensor function to BP incase we have false negatives to notify such as Walkers Overlaps events. We will need to look further into this logic. 

Fixes #7025 #7394    

#### Where has this been tested?

  * **Platform(s):** Ubuntu 20.04
  * **Python version(s):** .3.8.10
  * **Unreal Engine version(s):** 4.26

#### Possible Drawbacks

Normal Impulse not properly added in recorder. False positives appear.

### CODE REVIEW

File: Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp [hunk 1]
```diff
@@ -32,27 +35,37 @@ void ACollisionSensor::SetOwner(AActor *NewOwner)
   Super::SetOwner(NewOwner);
 
   /// @todo Deregister previous owner if there was one.
-
   if (NewOwner != nullptr)
```
Comment (xavisolesoft) [inline Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp:38]: **Code Simplification Suggestion**
This null check if could be removed as we are already checking now if the pointer is valid below.

Comment (Blyron) [reply Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp:38]: You are right, I'll do it!

File: Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp [hunk 2]
```diff
@@ -62,6 +75,29 @@ void ACollisionSensor::OnCollisionEvent(
...
       CollisionRegistry.end());
+}
+
+void ACollisionSensor::OnActorCollisionEvent(
+    AActor *Actor,
+    AActor *OtherActor,
+    FVector NormalImpulse,
+    const FHitResult &Hit)
+{
+  if (OtherActor == nullptr)
```
Comment (xavisolesoft) [inline Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp:86]:
```suggestion
  if (IsValid(OtherActor))
```

File: Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp [hunk 3]
```diff
@@ -62,6 +75,29 @@ void ACollisionSensor::OnCollisionEvent(
...
+  {
+    UE_LOG(LogCarla, Error, TEXT("ACollisionSensor::OnActorCollisionEvent Error with collided actor; Not valid.\n Collider actor %s"),
+      *(Actor->GetName()) );
+    return;
+  }
+
+  if (Actor == nullptr)
```
Comment (xavisolesoft) [inline Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp:93]:
```suggestion
  if (IsValid(Actor == nullptr))
```

File: Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp [hunk 4]
```diff
@@ -62,6 +75,29 @@ void ACollisionSensor::OnCollisionEvent(
...
+  {
+    UE_LOG(LogCarla, Error, TEXT("ACollisionSensor::OnActorCollisionEvent Error with collider actor; Not valid.\n Collided actor %s"),
+      *(OtherActor->GetName()) );
```
Comment (xavisolesoft) [inline Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp:96]:
```suggestion
      *(Other->GetName()) );
```

File: Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp [hunk 5]
```diff
@@ -62,6 +75,29 @@ void ACollisionSensor::OnCollisionEvent(
...
+void ACollisionSensor::OnActorCollisionEvent(
+    AActor *Actor,
+    AActor *OtherActor,
+    FVector NormalImpulse,
+    const FHitResult &Hit)
+{
+  if (OtherActor == nullptr)
+  {
+    UE_LOG(LogCarla, Error, TEXT("ACollisionSensor::OnActorCollisionEvent Error with collided actor; Not valid.\n Collider actor %s"),
+      *(Actor->GetName()) );
```
Comment (xavisolesoft) [inline Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp:89]:
```suggestion
      *(OtherActor->GetName()) );
```

File: Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp [hunk 6]
```diff
@@ -74,17 +110,94 @@ void ACollisionSensor::OnCollisionEvent(
...
+      AActor* OtherActor,
+      UPrimitiveComponent* OtherComp,
+      FVector NormalImpulse,
+      const FHitResult& Hit)
+{
+  AActor* Actor = HitComp->GetOwner();
+  if (OtherActor == nullptr)
+  {
+    UE_LOG(LogCarla, Error, TEXT("ACollisionSensor::OnComponentCollisionEvent Error with collided actor; Not valid.\n Collider actor %s"),
+      *(Actor->GetName()) );
```
Comment (xavisolesoft) [inline Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp:163]:
```suggestion
      *(OtherActor->GetName()) );
```

File: Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp [hunk 7]
```diff
@@ -74,17 +110,94 @@ void ACollisionSensor::OnCollisionEvent(
...
+    return;
+  }
+
+  if (Actor == nullptr)
+  {
+    UE_LOG(LogCarla, Error, TEXT("ACollisionSensor::OnComponentCollisionEvent Error with collider actor; Not valid.\n Collided actor %s"),
+      *(OtherActor->GetName()) );
```
Comment (xavisolesoft) [inline Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp:170]:
```suggestion
      *(Actor->GetName()) );
```

File: Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp [hunk 8]
```diff
@@ -74,17 +110,94 @@ void ACollisionSensor::OnCollisionEvent(
...
+
+void ACollisionSensor::OnComponentCollisionEvent(
+      UPrimitiveComponent* HitComp,
+      AActor* OtherActor,
+      UPrimitiveComponent* OtherComp,
+      FVector NormalImpulse,
+      const FHitResult& Hit)
+{
+  AActor* Actor = HitComp->GetOwner();
+  if (OtherActor == nullptr)
```
Comment (xavisolesoft) [inline Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp:160]:
```suggestion
  if (IsValid(OtherActor))
```

File: Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp [hunk 9]
```diff
@@ -74,17 +110,94 @@ void ACollisionSensor::OnCollisionEvent(
...
+  {
+    UE_LOG(LogCarla, Error, TEXT("ACollisionSensor::OnComponentCollisionEvent Error with collided actor; Not valid.\n Collider actor %s"),
+      *(Actor->GetName()) );
+    return;
+  }
+
+  if (Actor == nullptr)
```
Comment (xavisolesoft) [inline Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp:167]:
```suggestion
  if (IsValid(Actor))
```

File: Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp [hunk 10]
```diff
@@ -74,17 +110,94 @@ void ACollisionSensor::OnCollisionEvent(
     }
   }
 
-  const auto &Episode = GetEpisode();
+  const auto& CurrentEpisode = GetEpisode();
+  constexpr float TO_METERS = 1e-2;
```
Comment (xavisolesoft) [inline Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp:114]: **Prevent code logic duplication**
I suggest to create a common function for all collision events in CollisionSensor as the logic is the same. The structure should look like this:

```c++

//This one do not need to be an UFunction I think
void ACollisionSensor::OnCollisionEvent(
    AActor *Actor,
    AActor *OtherActor,
    FVector NormalImpulse,
    const FHitResult &Hit)
{
  //Paste here the shared code
}

void ACollisionSensor::OnActorCollisionEvent(
    AActor *Actor,
    AActor *OtherActor,
    FVector NormalImpulse,
    const FHitResult &Hit)
{
  OnCollisionEvent(Actor, OtherActor, ormalImpulse, Hit);
}

void ACollisionSensor::OnComponentCollisionEvent(
      UPrimitiveComponent* HitComp,
      AActor* OtherActor,
      UPrimitiveComponent* OtherComp,
      FVector NormalImpulse,
      const FHitResult& Hit)
{
  AActor* Actor = HitComp->GetOwner();
  OnCollisionEvent(Actor, OtherActor, ormalImpulse, Hit);
 }
```

**Review by xavisolesoft — CHANGES_REQUESTED** [review]

File: Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp (hunk 10 above)
Comment (Blyron) [reply Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp:114]: Applied

File: Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp (hunk 9 above)
Comment (Blyron) [reply Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp]: Applied

File: Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp (hunk 8 above)
Comment (Blyron) [reply Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp]: Applied

File: Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp (hunk 3 above)
Comment (Blyron) [reply Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp]: Applied

File: Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp (hunk 2 above)
Comment (Blyron) [reply Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp]: Applied

File: Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp (hunk 1 above)
Comment (Blyron) [reply Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Sensor/CollisionSensor.cpp]: Applied

**Review by xavisolesoft — APPROVED** [review]

**Pull request merged**